### ISSUE TITLE
Mean Sigmoid Cross Entropy Loss question

### ISSUE THREAD

**Issue by djo-koconi** [anicolson/DeepXi#21]
Hi Aaron.

Could you please help me understand the mean_sigmoid_cross_entropy loss implementation in DeepXi?

So the inputs are noisy speech magnitude spectrum mbatch[0], and predicted SNR net.target_ph: mbatch[1] .

I am having difficulties figuring it out from the code. How are you calculating the loss between the features and the SNR prediction (as they seem to be different types of information and even shape) ?

**Comment by anicolson** [comment]
Could you please help me understand the mean_sigmoid_cross_entropy loss implementation in DeepXi?

Just a combination of tf.nn.sigmoid_cross_entropy_with_logits and tf.math.reduce_sum


So the inputs are noisy speech magnitude spectrum mbatch[0], and predicted SNR net.target_ph: mbatch[1] .

The inputs to the loss function are the outputs of the network and the mapped a priori SNR (net.target_ph: mbatch[1]). net.input_ph: mbatch[0] is the input to the network.


Paper mentions IRM, but I don't see it used.

​In the paper is a comparison between a deep learning-based a priori SNR estimator for MMSE approaches to speech enhancement to an IRM estimator. The square-root Weiner filter (which has the same form as the IRM) is available to use if you look at lib/dev/gain.py.
________________________________

**Comment by djo-koconi** [comment]
Thank you Aaron. 

> The inputs to the loss function are the outputs of the network and the mapped a priori SNR (net.target_ph: mbatch[1]). net.input_ph: mbatch[0] is the input to the network.

This confuses me as they do not have the same shape.
What the output of the network represents in this case?

My understanding is that mbatch keep: [0]- Noisy signal mag spectrum, [1] SNR predictions , [2] Sequence lengths. 


So is it correct to estimate mag spectrum of signal **after noise reduction** during training as:

`for i in range(mbatch_size_iter):`
    `    curent_sample_len=mbatch[2][i]`
    `    noisy_file_MAG=mbatch[0][i][0:curent_sample_len]`
    `    snr_prediction=mbatch[1][position:position+curent_sample_len]`
` `
    `    enhansed_sample_MAG= np.multiply(noisy_file_MAG, gain.gfunc(snr_prediction, snr_prediction+1, gtype='mmse-lsa'))`

**Comment by anicolson** [comment]
The target sequences are concatenated together

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by anicolson** [comment]
Look at boolean_mask in ResNet.py

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by s-tomic** [comment]
I understand that they are concatenated, so  I load them according to the original seq length.
Just I'm not sure what exactly it represent. Do I need to undo the masking before using the output?

Please tell me if you can. do I have the right understanding what mbatch outputs represent. Is the way of utilizing mbatch to find enhanced signal during training the right way?

**Comment by djo-koconi** [comment]
OK, thank you.
I see that it is masked according to the seq length, but what does the ResNet output actually represent?

> The inputs to the loss function are the outputs of the network and the mapped a priori SNR (net.target_ph: mbatch[1]). net.input_ph: mbatch[0] is the input to the network.

You meant **inputs** of the network and the mapped a priori SNR? Right?

So its calculating loss between the features (input) and SNR prediction (output) ?  I cant follow up the logic it seems.

**Comment by anicolson** [comment]
The loss is calculated between the output of a neural network and the labels/target. The target in  this case is the mapped a priori snr. The neural net learns to map the input to the target. Therefore, the output of the neural net in this case is the mapped a priori snr. The placeholders indicate the inputs to the graph. The output of the network is not one of the inputs to the graph. You need to provide the inputs to the placeholder, so that outputs from the network can be computed for the loss function. I suggest you familiarise yourself with how graphs work with tensorflow.

Hope this helps.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by djo-koconi** [comment]
Thank you Aaron. 

I'm working on it :)
